components/user-mgt/org.wso2.carbon.identity.user.pre.update.profile.action/src/main/java/org/wso2/carbon/identity/user/pre/update/profile/action/internal/execution/PreUpdateProfileResponseProcessor.java

```diff
@@ -44,6 +88,403 @@ public ActionExecutionStatus<Success> processSuccessResponse(FlowContext flowCon
+            if (matcher.group(2) != null) {
+                int valueIndex = Integer.parseInt(matcher.group(3));
+                String[] values = (String[]) claim.getValue();
+                if (valueIndex < 0 || valueIndex >= values.length) {
+                    throw new ActionExecutionResponseProcessorException("Invalid index for multivalued claim: "
+                            + valueIndex);
+                }
+            }
+        } else {
+            throw new ActionExecutionResponseProcessorException("Invalid path format: " + path);
+        }
+    }
+
+    private void validateUserValuesForAddAndModifyOperations(String claimValue, String claimUri, String userId,
+                                                             Operation operation)
+            throws ActionExecutionResponseProcessorException {
+
+        UniqueIDUserStoreManager userStoreManager = getUserStoreManager();
+        try {
+            Map<String, String> userClaimValues = userStoreManager.getUserClaimValuesWithID(userId,
+                    new String[] {claimUri}, UserCoreConstants.DEFAULT_PROFILE);
+            if (operation == Operation.ADD && userClaimValues.get(claimUri) != null && !claimValue.trim().isEmpty()) {
+                throw new ActionExecutionResponseProcessorException(claimUri + " claim value already exists for user.");
+            } else if (operation.equals(Operation.REPLACE) && userClaimValues.get(claimUri) == null) {
+                throw new ActionExecutionResponseProcessorException(claimUri + " claim value does not exist for user.");
+            }
+        } catch (org.wso2.carbon.user.core.UserStoreException e) {
+            throw new ActionExecutionResponseProcessorException("Failed to retrieve user claims from user store.", e);
+        }
+    }
+
+    private void validateImmutableClaims(String claimUri) throws ActionExecutionResponseProcessorException {
+
+        if (claimUri.equals(ID_CLAIM_URI) || claimUri.equals(CREATED_CLAIM_URI) ||
+                claimUri.equals(FAMILY_NAME_CLAIM_URI) || claimUri.contains(IDENTITY_CLAIM_URI_PREFIX)) {
+            throw new ActionExecutionResponseProcessorException("Immutable claims cannot be added or modified: " +
+                    claimUri);
+        }
+    }
+
+    private void validateFlowInitiatorClaims(String claimUri, Optional<LocalClaim> localClaim)
+            throws ActionExecutionResponseProcessorException {
+
+        if (localClaim.get().getFlowInitiator()) {
+            throw new ActionExecutionResponseProcessorException("Flow initiator claims cannot be added: " + claimUri);
+        }
+    }
+
+    private void validateGroupAndRoleClaims(String claimUri) 
+            throws ActionExecutionResponseProcessorException {
+
+        if (claimUri.equals(GROUP_CLAIM_URI) || claimUri.equals(ROLE_CLAIM_URI)) {
+            throw new ActionExecutionResponseProcessorException("Group/Role claims cannot be added: " + claimUri);
```

Comment:
Let's improve the message saying 'Groups/Roles are not allowed to be added' or "Adding Groups/Roles is not allowed"